fendermint/vm/topdown/src/disabled.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::{
+    BlockHash, BlockHeight, Error, IPCParentFinality, InMemoryFinalityProvider,
+    ParentFinalityProvider, ParentViewProvider,
+};
+use async_stm::{StmError, StmResult};
+use ipc_agent_sdk::message::ipc::ValidatorSet;
+use ipc_sdk::cross::CrossMsg;
+
+/// The parent finality provider could have all functionalities disabled.
+#[derive(Clone)]
+pub enum MaybeDisabledProvider {
+    Disabled,
+    Enabled(InMemoryFinalityProvider),
+}
+
+impl MaybeDisabledProvider {
+    pub fn disabled() -> Self {
+        Self::Disabled
+    }
+
+    pub fn enabled(inner: InMemoryFinalityProvider) -> Self {
+        Self::Enabled(inner)
+    }
+}
+
+macro_rules! disabled {
+    () => {
+        Err(StmError::Abort(Error::ProviderNotEnabled))
+    };
+}
+
+impl ParentViewProvider for MaybeDisabledProvider {
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>, Error> {
+        match self {
+            MaybeDisabledProvider::Disabled => disabled!(),
+            MaybeDisabledProvider::Enabled(p) => p.latest_height(),
+        }
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> StmResult<Option<BlockHash>, Error> {
+        match self {
+            MaybeDisabledProvider::Disabled => disabled!(),
+            MaybeDisabledProvider::Enabled(p) => p.block_hash(height),
+        }
+    }
+
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>, Error> {
+        match self {
+            MaybeDisabledProvider::Disabled => disabled!(),
+            MaybeDisabledProvider::Enabled(p) => p.validator_set(height),
+        }
+    }
+
+    fn top_down_msgs(&self, height: BlockHeight) -> StmResult<Vec<CrossMsg>, Error> {
+        match self {
+            MaybeDisabledProvider::Disabled => disabled!(),
+            MaybeDisabledProvider::Enabled(p) => p.top_down_msgs(height),
+        }
+    }
+
+    fn new_parent_view(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+        top_down_msgs: Vec<CrossMsg>,
+    ) -> StmResult<(), Error> {
+        match self {
+            MaybeDisabledProvider::Disabled => disabled!(),
+            MaybeDisabledProvider::Enabled(p) => {
+                p.new_parent_view(height, block_hash, validator_set, top_down_msgs)
+            }
+        }
+    }
+}
+
+impl ParentFinalityProvider for MaybeDisabledProvider {
+    fn last_committed_finality(&self) -> StmResult<IPCParentFinality, Error> {
+        match self {
+            MaybeDisabledProvider::Disabled => disabled!(),
+            MaybeDisabledProvider::Enabled(p) => p.last_committed_finality(),
+        }
+    }
+
+    fn next_proposal(&self) -> StmResult<Option<IPCParentFinality>, Error> {
```

Comment:
So you had two options when you implemented a `Toggle`: 
* return an error, like you do
* return an "empty" result, which can be slightly for each method

The problem with returning an error are:
* I assume you had to extend the error types with yet another thing the caller has to reason about 
* where you[ call it from](https://github.com/consensus-shipyard/fendermint/pull/231/files#diff-38169d6d71afc04b5bc631226485edb63af1309907ec52553d956d0c55e86259R101-R104) in the `ChainMessageInterpreter` during `prepare_proposal` in the previous PR it's not expected to return an error, and returning one will immiately crash Fendermint

This is why I am begging you to not return errors where an empty result would do, don't return errors for things that can be expected. Only return errors where they are unexpected if you plan to use `?` inside interpreters, because they are fatal errors. 

The condition of not having a parent is not fatal. You could pass to the interpreter an `Option<InMemoryFinalityProvider>` which would tell the interpreter that there isn't any parent, but instead it looks like a normal provider it can call, except it bombs. 

And even on the `Enabled` case, why should it be an error that the syncer doesn't yet have data? We don't _have_ to propose a finality, we can just handle `None` and move on. Yet this one returns an error and it kills the app, or forces everyone to think whether its' enabled or disabled. 

I think it should return empties: it doesn't have any data about any height, and never proposes anything new. You can also change the `check` function to return `true` or `false` instead of an error, and then disabled can always return `false`. The only question is what to return from `last_committed_finality` but I think that can even be made into an internal method. The only thing that can return an error is `on_committed_finality` because there should be no commits as there is not parent.

Reply:
Errors are internalized and returns only option.